hcl/hclsyntax/expression.go

```diff
@@ -473,20 +473,29 @@ func (e *ConditionalExpr) Value(ctx *hcl.EvalContext) (cty.Value, hcl.Diagnostic
 	falseResult, falseDiags := e.FalseResult.Value(ctx)
 	var diags hcl.Diagnostics
 
-	var resultType cty.Type
+	resultType := cty.DynamicPseudoType
 	convs := make([]convert.Conversion, 2)
 
 	switch {
 	// If either case is a dynamic null value (which would result from a
 	// literal null in the config), we know that it can convert to the expected
 	// type of the opposite case, and we don't need to speculatively reduce the
 	// final result type to DynamicPseudoType.
+
+	// If we know that either Type is a DynamicPseudoType, we can be certain
+	// that the other value can convert since it's a pass-through, and we don't
+	// need to unify the types. If the final evaluation results in the dynamic
+	// value being returned, there's no conversion we can do, so we return the
+	// value directly.
```

Comment:
I think this _is_ safe indeed, but I think it's for a slightly different reason than this comment states: the rules for unknown values and unknown types is that HCL can make any assumption that would lead to an error if it were violated once the unknown becomes known.

In this case, if we have `cond ? known : dynamicval` then it's safe to assume that the result will be compatible with `known` because once we _do_ know the type of that second operand, we'll fail at _that_ point if the types won't unify.

(Because we're using `UnifyUnsafe` it's still possible that the final result will be of a different type than we predicted, but it should always be one that can be safely/implicitly converted back to what we predicted and thus not cause any downstream misbehavior.)

Reply:
Maybe my comment wasn't worded as well as it could have, so we may be on the same page here. With `known : dynamicval`; the expression as a whole is resolves to `DynamicPseudoType`. If the condition is known, then we either return the `dynamicval`, which of course is the same type, or `known`, which converts to itself when the type is dynamic. Hence there's no need to unify.

Reply:
Maybe to put it another way: the cases this directly solves don't need to assume the result will be compatible with `known`, because `dynamicval` will never be known or evaluated. 

In cases where `dynamicval` is known later on, "later on' is an entirely new evaluation, where we have more type information and will return a more precise type using the original `UnifyUnsafe` code path.